File: src/ds/dsu.ts
import { multiArray } from '../util/array';

class DSU {
  p: number[];
  constructor(size: number) {
    this.p = multiArray(-1, size + 1);
  }
  /**
   *
   * @param current
   * @returns The index of the leader of that current node's group
   */
  group(current: number): number {
    if (this.p[current] < 0) {
      return current;
    } else {
      this.p[current] = this.group(this.p[current]);
      return this.p[current];
    }
  }
  /**
   *
   * @param a
   * @param b
   * @returns If node a and b successfully joined into the same group
   */
  join(a: number, b: number): boolean {
    a = this.group(a);
    b = this.group(b);
    if (a == b) {
      return false;
    }
    if (this.p[a] > this.p[b]) {
      [a, b] = [b, a];
    }
    this.p[a] += this.p[b];
    this.p[b] = a;
    return true;
  }
  /**
   *
   * @param current
   * @returns The size of the current node's group
   */
  size(current: number): number {
    current = this.group(current);
    return -this.p[current];
  }
}

export { DSU };

File: src/lib/dsu.cpp
struct DSU {
  vector<int> p;
  DSU() { }
  void init(int _) {
    p.assign(_ + 1, -1);
  }
  DSU(int _) {
    init(_);
  }
  int par(int u) { 
    return p[u] < 0 ? u : (p[u] = par(p[u])); 
  }
  bool same(int u, int v) { 
    return par(u) == par(v); 
  }
  int size(int u) { 
    return - p[par(u)]; 
  }
  bool join(int u, int v) {
    if ((u = par(u)) == (v = par(v))) 
      return false;
    if (p[u] > p[v]) 
      swap(u, v);
    p[u] += p[v]; p[v] = u;
    return true;
  }
};